File: apps/desktop/src/lib/analytics.ts
import { invoke } from "@tauri-apps/api/core";

import { setSetting } from "./settings";
import { SETTING_KEYS } from "../types";

export type AnalyticsInfo = {
  /** False in local dev and in any build without an ingest endpoint. */
  configured: boolean;
  enabled: boolean;
  /** The user's own install identifier, so they can quote it when asking for
   * their records to be deleted. Null while opted out. The ingest endpoint is
   * deliberately not exposed. */
  installId: string | null;
};

export function getAnalyticsConfig(): Promise<AnalyticsInfo> {
  return invoke<AnalyticsInfo>("analytics_config");
}

/**
 * Records the consent choice in both places it has to live: the settings table
 * (which survives a restart) and the running process (which sends the events).
 *
 * The two writes are ordered so a partial failure always fails toward NOT
 * sending. Enabling persists first, so a failed runtime call leaves the app
 * quiet until the next launch. Disabling stops the runtime first, so a failed
 * persist has already taken effect. Never the reverse.
 *
 * `analytics_set_enabled` also owns the install identifier: the backend mints
 * it on opt-in and deletes it on opt-out, so it never exists for someone who
 * has not agreed.
 */
export async function applyAnalyticsConsent(enabled: boolean): Promise<void> {
  if (enabled) {
    await setSetting(SETTING_KEYS.analytics, true);
    await setAnalyticsEnabled(true);
    return;
  }
  await setAnalyticsEnabled(false);
  await setSetting(SETTING_KEYS.analytics, false);
}

function setAnalyticsEnabled(enabled: boolean): Promise<void> {
  return invoke<void>("analytics_set_enabled", { enabled });
}

File: apps/desktop/src-tauri/src/commands/analytics.rs
use serde::Serialize;
use serde_json::Value;
use tauri::State;

use crate::analytics;
use crate::errors::Result;
use crate::storage::settings;
use crate::AppState;

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AnalyticsInfo {
    /// Whether this build carries an analytics endpoint at all. False in local
    /// dev and in any fork, which is what keeps the consent prompt from
    /// appearing where it would have nothing to talk to.
    pub configured: bool,
    pub enabled: bool,
    /// The install identifier, shown in Settings so a user can quote it when
    /// asking for their analytics records to be deleted. `None` while opted
    /// out, because it does not exist then. The ingest endpoint is
    /// deliberately NOT exposed here.
    pub install_id: Option<String>,
}

#[tauri::command]
pub fn analytics_config() -> AnalyticsInfo {
    let analytics = analytics::handle();
    AnalyticsInfo {
        configured: analytics.is_configured(),
        enabled: analytics.is_enabled(),
        install_id: analytics.install_id(),
    }
}

/// Applies the consent choice to the running process, and owns the install
/// identifier's whole lifetime.
///
/// The identifier is minted here rather than in the frontend so it is never
/// created for someone who has not opted in, and it is deleted on opt-out —
/// so turning analytics off genuinely forgets the device rather than pausing
/// it, and turning it back on starts a new identity.
///
/// Persisting the consent flag itself stays with the frontend (`settings_set`)
/// so the settings cache stays coherent and the two writes can be ordered to
/// fail toward not sending.
#[tauri::command]
pub async fn analytics_set_enabled(app_state: State<'_, AppState>, enabled: bool) -> Result<()> {
    let analytics = analytics::handle();
    if enabled {
        let existing = settings::all(&app_state.pool)
            .await?
            .get(analytics::INSTALL_ID_SETTING_KEY)
            .and_then(Value::as_str)
            .map(str::to_string);
        let install_id = match existing {
            Some(id) => id,
            None => {
                let id = analytics::new_install_id();
                settings::set(
                    &app_state.pool,
                    analytics::INSTALL_ID_SETTING_KEY,
                    &Value::String(id.clone()),
                )
                .await?;
                id
            }
        };
        analytics.set_install_id(Some(install_id));
        analytics.set_enabled(true);
    } else {
        analytics.set_enabled(false);
        analytics.set_install_id(None);
        settings::delete(&app_state.pool, analytics::INSTALL_ID_SETTING_KEY).await?;
    }
    Ok(())
}
